### PR TITLE
知識王和1A2B小遊戲

### PR DESCRIPTION
**Pull request by wow890209**
## Why need this change? / Root cause: 
-修改discordkt部分
-增加1A2B小遊戲
## Changes made:
-discordkt部分註解掉，正在用jda替換
-1A2B遊戲邏輯和流程
## Test Scope / Change impact:
-測試1A2B的判斷規則

### CODE REVIEW

File: GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/domain/GuessNum1A2B.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        var b = 0
+        for (i in 0 until 4) {
+            if (number[i] == secretNumber[i]) {
+                a++
+            } else if (secretNumber.contains(number[i])) {
+                b++
+            }
+        }
+        return "${a}A${b}B"
+    }
```
Comment (akechetaki10736) [inline GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/domain/GuessNum1A2B.kt:21]: Please format after you finish.
Ctrl - Alt - L

File: main/pom.xml [hunk 2]
```diff
@@ -24,6 +24,14 @@
             <groupId>tw.waterballsa.utopia</groupId>
             <artifactId>landing-experience</artifactId>
         </dependency>
+        <dependency>
+            <groupId>tw.waterballsa.utopia</groupId>
+            <artifactId>knowledgeking</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>tw.waterballsa.utopia</groupId>
+            <artifactId>GuessNum1A2B</artifactId>
```
Comment (Johnny850807) [inline main/pom.xml:33]: 小寫駱駝命名 guessNum1A2B

File: pom.xml [hunk 3]
```diff
@@ -15,18 +15,15 @@
...
         <module>discord-impl-jda</module>
         <module>commons</module>
-        <!--        <module>knowledge-king</module>-->
-        <!--        <module>gamification</module>-->
+<!--        <module>knowledge-king</module>-->
+<!--        <module>gamification</module>-->
         <module>landing-experience</module>
-        <module>chatgpt-api</module>
-        <module>forum-experience</module>
-        <module>gaas</module>
```
Comment (Johnny850807) [inline pom.xml:23]: 你是不是刪到？

File: pom.xml [hunk 4]
```diff
@@ -123,11 +79,6 @@
                 <artifactId>JDA</artifactId>
                 <version>${JDA.version}</version>
             </dependency>
-            <dependency>
-                <groupId>com.fasterxml.jackson.core</groupId>
-                <artifactId>jackson-databind</artifactId>
-                <version>${jackson.version}</version>
-            </dependency>
```
Comment (Johnny850807) [inline pom.xml:130]: 你好像刪到很多東西

File: pom.xml [hunk 5]
```diff
@@ -163,21 +114,6 @@
...
-            <dependency>
-                <groupId>tw.waterballsa.utopia</groupId>
-                <artifactId>chatgpt-api</artifactId>
-                <version>${revision}</version>
-            </dependency>
-            <dependency>
-                <groupId>tw.waterballsa.utopia</groupId>
-                <artifactId>gaas</artifactId>
-                <version>${revision}</version>
-            </dependency>
```
Comment (Johnny850807) [inline pom.xml:180]: 你好像刪到很多東西

File: pom.xml [hunk 6]
```diff
@@ -90,30 +70,6 @@
...
-            <version>${javax.annotation-api.version}</version>
-        </dependency>
-
-        <dependency>
-            <groupId>com.github.MinnDevelopment</groupId>
-            <artifactId>jda-ktx</artifactId>
-            <version>${jda-ktx.version}</version>
-        </dependency>
-
     </dependencies>
```
Comment (Johnny850807) [inline pom.xml:73]: 你好像刪到很多東西

File: pom.xml [hunk 7]
```diff
@@ -37,25 +34,8 @@
...
-            <url>https://repo1.maven.org/maven2</url>
-            <layout>default</layout>
-        </repository>
-        <repository>
-            <id>jitpack.io</id>
-            <url>https://jitpack.io</url>
-        </repository>
-    </repositories>
-
     <dependencies>
```
Comment (Johnny850807) [inline pom.xml:39]: 你好像刪到很多東西

File: GuessNum1A2B/src/test/resources/junit-platform.properties.properties [hunk 8]
```diff
@@ -0,0 +1 @@
+junit.jupiter.testinstance.lifecycle.default = per_class
```
Comment (Johnny850807) [inline GuessNum1A2B/src/test/resources/junit-platform.properties.properties:1]: 這是啥？

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
你好像刪到很多東西、CI 還沒過，感覺版本控制有一些問題，你注意一下

File: GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/listener.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import net.dv8tion.jda.api.JDA
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent
+import net.dv8tion.jda.api.events.session.ReadyEvent
+import tw.waterballsa.utopia.GuessNum1A2B.domain.GuessNum1A2B
+import tw.waterballsa.utopia.GuessNum1A2B.domain.generateSecretNumber
+import tw.waterballsa.utopia.commons.config.WsaDiscordProperties
+import tw.waterballsa.utopia.jda.listener
+
+val logger = KotlinLogging.logger{}
+val GuessNumChannelID = 1093177540346130453
```
Comment (akechetaki10736) [inline GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/listener.kt:14]: Why is the first letter of the variable name capitalized ? 
Please follow the naming convention. 
Also, this should be placed in the yml file.

File: GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/listener.kt [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+var game = GuessNum1A2B(generateSecretNumber())
+
+
+fun GuessNum1A2BListerner(wsa: WsaDiscordProperties, jda: JDA) = listener {
```
Comment (akechetaki10736) [inline GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/listener.kt:18]: ditto

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt [hunk 11]
```diff
@@ -1,43 +1,43 @@
...
+import dev.kord.core.behavior.channel.MessageChannelBehavior
+import dev.kord.core.event.gateway.ReadyEvent
+import kotlinx.coroutines.delay
+import me.jakejmattson.discordkt.dsl.listeners
+
+var game : Game? = null
+fun Listener() = listeners {
+
+    on<ReadyEvent> {
+        val c = MessageChannelBehavior(Snowflake(1038657903437037590),discord.kord).asChannel();
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt:13]: what is "c" meaning ?
Please clarify your variable name.

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt [hunk 12]
```diff
@@ -1,43 +1,43 @@
...
-//
-//}
+
+import dev.kord.common.entity.Snowflake
+import dev.kord.core.behavior.channel.MessageChannelBehavior
+import dev.kord.core.event.gateway.ReadyEvent
+import kotlinx.coroutines.delay
+import me.jakejmattson.discordkt.dsl.listeners
+
+var game : Game? = null
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt:9]: `lateinit var game: Game`

File: pom.xml [hunk 13]
```diff
@@ -15,18 +15,15 @@
...
-        <module>chatgpt-api</module>
-        <module>forum-experience</module>
-        <module>gaas</module>
     </modules>
 
     <properties>
         <revision>1.0-SNAPSHOT</revision>
         <kotlin.version>1.8.10</kotlin.version>
-        <kotlinx-coroutines-core.version>1.6.4</kotlinx-coroutines-core.version>
+        <kotlinx-coroutines-core.version>1.6.1</kotlinx-coroutines-core.version>
```
Comment (akechetaki10736) [inline pom.xml:26]: why downgrade ?

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt [hunk 14]
```diff
@@ -1,43 +1,43 @@
...
+            game = Game(c)
+
+            game!!.CreateRanking()
+            val c = game!!.CreateMenu(discord)
+            c.createMessage("知識王比賽已準備好 10秒後開始答題")
+            delay(10000)
+
+            while (!game!!.IsEndGame()) {
+                c.createMessage("開始下一題")
+                game!!.Next()
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt:32]: ditto

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt [hunk 15]
```diff
@@ -1,43 +1,43 @@
...
+                delay(10500)
+                c.createMessage("答題結束 等待兩秒開始")
+                game!!.UpdateRanking()
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/listeners.kt:35]: ditto

File: GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/domain/GuessNum1A2B.kt [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package tw.waterballsa.utopia.GuessNum1A2B.domain
+
+import java.util.*
+
+class GuessNum1A2B(secret : String){
+    private val secretNumber = secret
```
Comment (akechetaki10736) [inline GuessNum1A2B/src/main/kotlin/tw/waterballsa/utopia/GuessNum1A2B/domain/GuessNum1A2B.kt:6]: `class GuessNum1A2B(private val secret: String)`

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt [hunk 17]
```diff
@@ -1,217 +1,217 @@
...
+    private val kingoftheQuiz = KingoftheQuiz()
+    private val channel = _channel
+    private var solutions : List<Solution>? = null
+    private var current = -1
+    private var playerRank = mutableMapOf<String, Int>()
+    suspend fun CreateMenu(dc: Discord) : MessageChannelBehavior
+    {
+        solutions = GooleSheet().ReadSolution()
+        gamemessage = channel.createMessage("loading")
+        val b = dc.kord.rest.channel.startThreadWithMessage(
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt:76]: what is "b" and why we still have kord here ?

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt [hunk 18]
```diff
@@ -1,217 +1,217 @@
...
+            channel.id,
+            gamemessage!!.id,
+            name = "123",
+            ArchiveDuration.Day
+        ) {}.id
+
+        val m = MessageChannelBehavior(b, dc.kord)
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt:83]: ditto

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt [hunk 19]
```diff
@@ -1,217 +1,217 @@
...
+class Game(_channel : MessageChannel)
+{
+    private var gamemenu : Menu? = null
+    private var gamerank : Message? = null
+    private var gamemessage : Message? = null
+    private val kingoftheQuiz = KingoftheQuiz()
+    private val channel = _channel
+    private var solutions : List<Solution>? = null
+    private var current = -1
+    private var playerRank = mutableMapOf<String, Int>()
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt:71]: Please consider `lateinit`

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt [hunk 20]
```diff
@@ -1,217 +1,217 @@
...
+
+    suspend fun CreateRanking()
+    {
+        gamerank = channel.createMessage("Rank:")
+    }
+
+    suspend fun UpdateRanking()
+    {
+        if (current == -1) return
+        val playerdata = kingoftheQuiz.GetAnswerList()
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt:140]: Naming convention

File: knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt [hunk 21]
```diff
@@ -1,217 +1,217 @@
...
+    {
+        return current+1 == solutions!!.size
+    }
+    suspend fun Next()
+    {
+        gamemenu!!.nextPage()
+        gamemessage!!.edit(gamemenu!!)
+        kingoftheQuiz.Prepare(solutions!![++current], 10)
+    }
+}
```
Comment (akechetaki10736) [inline knowledge-king/src/main/kotlin/tw/waterballsa/utopia/knowledgeking/KinoftheQuiz.kt:191]: Format your code

**Review by akechetaki10736 — CHANGES_REQUESTED** [review]
Please seriously consider naming variables, convention and format.

**Pull request closed without merging**